rosbag2_transport/test/rosbag2_transport/test_play_services.cpp

```diff
@@ -0,0 +1,236 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+#include <utility>
+
+#include "rclcpp/rclcpp.hpp"
+
+#include "rosbag2_interfaces/srv/is_paused.hpp"
+#include "rosbag2_interfaces/srv/pause.hpp"
+#include "rosbag2_interfaces/srv/resume.hpp"
+#include "rosbag2_interfaces/srv/toggle_paused.hpp"
+#include "rosbag2_transport/player.hpp"
+#include "test_msgs/msg/basic_types.hpp"
+#include "test_msgs/message_fixtures.hpp"
+
+#include "rosbag2_play_test_fixture.hpp"
+
+using namespace ::testing;  // NOLINT
+
+class PlaySrvsTest : public RosBag2PlayTestFixture
+{
+public:
+  using Pause = rosbag2_interfaces::srv::Pause;
+  using Resume = rosbag2_interfaces::srv::Resume;
+  using TogglePaused = rosbag2_interfaces::srv::TogglePaused;
+  using IsPaused = rosbag2_interfaces::srv::IsPaused;
+
+  PlaySrvsTest()
+  : RosBag2PlayTestFixture(),
+    client_node_(std::make_shared<rclcpp::Node>("test_play_client"))
+  {}
+
+  virtual ~PlaySrvsTest()
+  {
+    exec_.cancel();
+    rclcpp::shutdown();
+    spin_thread_.join();
+    play_thread_.join();
+  }
+
+  /// Use SetUp instead of ctor because we want to ASSERT some preconditions for the tests
+  void SetUp() override
+  {
+    setup_player();
+
+    const std::string ns = "/" + player_name_;
+    cli_pause_ = client_node_->create_client<Pause>(ns + "/pause");
+    cli_resume_ = client_node_->create_client<Resume>(ns + "/resume");
+    cli_toggle_paused_ = client_node_->create_client<TogglePaused>(ns + "/toggle_paused");
+    cli_is_paused_ = client_node_->create_client<IsPaused>(ns + "/is_paused");
+    topic_sub_ = client_node_->create_subscription<test_msgs::msg::BasicTypes>(
+      test_topic_, 10,
+      std::bind(&PlaySrvsTest::topic_callback, this, std::placeholders::_1));
+
+    exec_.add_node(player_);
+    exec_.add_node(client_node_);
+    spin_thread_ = std::thread(
+      [this]() {
+        exec_.spin();
+      });
+
+    // Make sure all expected services are present before starting any test
```

Comment:
Does it make sense to offload this into a `service_manager` similar to `publication` and `subscription` manager?
Not a show-stopper, just food for thought.

Reply:
I think that would make sense to me as a followup refactor if we want to use this pattern in other test fixtures. I like to avoid premature architecture.